In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('insurance.csv')


<br>**EDA**

In [ ]:
df.head()

age  sex     bmi  steps  children  smoker  region      charges  \
0   19    0  27.900   3009         0       1       3  16884.92400   
1   18    1  33.770   3008         1       0       2   1725.55230   
2   28    1  33.000   3009         3       0       2   4449.46200   
3   33    1  22.705  10009         0       0       1  21984.47061   
4   32    1  28.880   8010         0       0       1   3866.85520   

   insuranceclaim  
0               1  
1               1  
2               0  
3               0  
4               1

Setelah data diload saya menggunakan head function untuk melihat 5 data teratas

In [ ]:
print('Dimensity:', df.shape)
print(df.describe())

Dimensity: (1338, 9)
               age          sex          bmi         steps     children  \
count  1338.000000  1338.000000  1338.000000   1338.000000  1338.000000   
mean     39.207025     0.505232    30.663397   5328.623318     1.094918   
std      14.049960     0.500160     6.098187   2453.643236     1.205493   
min      18.000000     0.000000    15.960000   3000.000000     0.000000   
25%      27.000000     0.000000    26.296250   3008.000000     0.000000   
50%      39.000000     1.000000    30.400000   4007.000000     1.000000   
75%      51.000000     1.000000    34.693750   8004.000000     2.000000   
max      64.000000     1.000000    53.130000  10010.000000     5.000000   

            smoker       region       charges  insuranceclaim  
count  1338.000000  1338.000000   1338.000000     1338.000000  
mean      0.204783     1.515695  13270.422265        0.585202  
std       0.403694     1.104885  12110.011237        0.492871  
min       0.000000     0.000000   1121.873900  

Diatas adalah dimensi dan deskriptif statistik dari dataset

  DALAM MELAKUKAN EKSPLORASI DATA PADA KASUS INI ADA BEBERAPA HAL YANG SAYA PERHATIKAN TERLEBIH DAHULU:
- Missing value
- Categorical variable
- Scalling
- Imbalance target<br><br>
Jika ditemukan kondisi-kondisi diatas maka diperlukan data preprocessing untuk menghandle kondisi tersebut.

In [ ]:
#1 Missing value
print(df.isna().sum())

age               0
sex               0
bmi               0
steps             0
children          0
smoker            0
region            0
charges           0
insuranceclaim    0
dtype: int64


Karena tidak terdapat missing value jadi data sudah

In [ ]:
#2 Categorical variable
#Hal ini dilakukan agar nilai kategori bisa diubah menjadi nilai numerik yang bisa digunakan dalam pembuatan model.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             1338 non-null   int64  
 1   sex             1338 non-null   int64  
 2   bmi             1338 non-null   float64
 3   steps           1338 non-null   int64  
 4   children        1338 non-null   int64  
 5   smoker          1338 non-null   int64  
 6   region          1338 non-null   int64  
 7   charges         1338 non-null   float64
 8   insuranceclaim  1338 non-null   int64  
dtypes: float64(2), int64(7)
memory usage: 94.2 KB


Pada dataset ini seluruh feature/variabel memiliki tipe numerik sehingga kita tidak perlu melakukan encoding untuk feature.

In [ ]:
#3. Scalling
# rentang dan besaran yang berbeda yang dapat mempengaruhi proses pembelajaran

from sklearn.preprocessing import MinMaxScaler

# create scaler
scaler = MinMaxScaler()

# select numerical columns
num_cols = ["age","bmi","steps","children","charges"]

# fit and transform data
df[num_cols] = scaler.fit_transform(df[num_cols])


Karena vaiabel diatas memiliki rentangan yang berbeda maka saya lakukan scaling normalization dengan technique minmax agar hasil scale yang diberikan tidak ada negatif/(0-1)

In [ ]:
#4 Imbalance target
# that one class has more records than the other.
print(df["insuranceclaim"].value_counts())
print(df["insuranceclaim"].value_counts(normalize=True))

# Check the mean of each numerical column for each group of insurance claim
print(df.groupby("insuranceclaim").mean())

# Check the correlation matrix of the numerical columns
print(df.corr())

1    783
0    555
Name: insuranceclaim, dtype: int64
1    0.585202
0    0.414798
Name: insuranceclaim, dtype: float64
                     age       sex       bmi     steps  children    smoker  \
insuranceclaim                                                               
0               0.419781  0.486486  0.320731  0.506532  0.336216  0.045045   
1               0.490255  0.518519  0.448619  0.208607  0.135888  0.318008   

                  region   charges  
insuranceclaim                      
0               1.488288  0.122901  
1               1.535121  0.244252  
                     age       sex       bmi     steps  children    smoker  \
age             1.000000 -0.020856  0.109272 -0.167957  0.042469 -0.025019   
sex            -0.020856  1.000000  0.046371 -0.039470  0.017163  0.076185   
bmi             0.109272  0.046371  1.000000 -0.681149  0.012759  0.003750   
steps          -0.167957 -0.039470 -0.681149  1.000000  0.055346 -0.267845   
children        0.042469  0.017

Dengan melihat hasil diatas saya mendapat informasi bahwa dataset termasuk kedalam kategori imbalance yang dimana hal ini juga dapat mempengaruhi performa model, beberapa cara untuk mengatasi hal ini adalah dengan resampling dan menggunakan algoritma yang berbeda. Namun pada kasus ini saya tidak akan menerapkannya karena dataset sudah fix dan algoritma yang digunakan juga fix yaitu neural network.

In [ ]:
df.head()

age  sex       bmi     steps  children  smoker  region   charges  \
0  0.021739    0  0.321227  0.001284       0.0       1       3  0.251611   
1  0.000000    1  0.479150  0.001141       0.2       0       2  0.009636   
2  0.217391    1  0.458434  0.001284       0.6       0       2  0.053115   
3  0.326087    1  0.181464  0.999857       0.0       0       1  0.333010   
4  0.304348    1  0.347592  0.714693       0.0       0       1  0.043816   

   insuranceclaim  
0               1  
1               1  
2               0  
3               0  
4               1

#

Berikut adalah preview dari dataset setalah dilakukan data preprocessing.

In [ ]:
# Split the data into features and target
df_input = df.drop("insuranceclaim", axis=1)
df_target = df[["insuranceclaim"]]

Proses diatas adalah pemisahan data yang akan digunakan sebagai input dan target, dimana target disini adalah variabel insuranceclaim. Dan untuk feature/variabel yang digunakan sebagai input adalah semua variabel di dataset selain variabel insuranceclaim.

In [ ]:
#preview data input
df_input.head()

age  sex       bmi     steps  children  smoker  region   charges
0  0.021739    0  0.321227  0.001284       0.0       1       3  0.251611
1  0.000000    1  0.479150  0.001141       0.2       0       2  0.009636
2  0.217391    1  0.458434  0.001284       0.6       0       2  0.053115
3  0.326087    1  0.181464  0.999857       0.0       0       1  0.333010
4  0.304348    1  0.347592  0.714693       0.0       0       1  0.043816

In [ ]:
#preview data target
df_target.head()

insuranceclaim
0               1
1               1
2               0
3               0
4               1

In [ ]:
from sklearn.model_selection import train_test_split

# split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(df_input, df_target, test_size=0.2, random_state=42)

# split test set into test and validation sets
X_test, X_val, y_test, y_val = train_test_split(X_test, y_test, test_size=0.5, random_state=42)

#print the lenght after spliting
print("Train set size:", len(X_train))
print("Test set size:", len(X_test))
print("Validation set size:", len(X_val))

Train set size: 1070
Test set size: 134
Validation set size: 134


Proses diatas adalah proses spliting data denngan ketentuan 80% training data, 10% validation, 10%test

In [ ]:
#melihat dimensi dari input
df_input.shape

(1338, 8)

In [ ]:
!pip install tensorflow

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


Pembuatan model ANN dari no 1c

In [ ]:
import keras
from tensorflow.keras.models import Sequential
import tensorflow as tf

#number of feature
n = 8

num_class = 1

model1c = keras.models.Sequential(name="my_custom_model_1c")
model1c.add(tf.keras.Input(shape=(n,)))
model1c.add(keras.layers.core.Dense(2*n, activation='relu', name="Hiddenlayer"))
model1c.add(keras.layers.core.Dense(2*n,   activation='relu', name="Hiddenlayer2"))
model1c.add(keras.layers.core.Dense(num_class,   activation='sigmoid', name="Outputlayer"))
model1c.compile(optimizer='sgd', loss='binary_crossentropy', metrics=['accuracy'])

print(model1c.summary())

Model: "my_custom_model_1c"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 Hiddenlayer (Dense)         (None, 16)                144       
                                                                 
 Hiddenlayer2 (Dense)        (None, 16)                272       
                                                                 
 Outputlayer (Dense)         (None, 1)                 17        
                                                                 
Total params: 433
Trainable params: 433
Non-trainable params: 0
_________________________________________________________________
None


Berikut adalah pembuatan model ANN. Model terdiri dari 2 hiden layer dan 1 input&output layer. Ketentuan input dan neuron:
- Pada layer input = n (jumlah feature) = 8
- hiden layer1 = 2*n
- hiden layer2 = 2*n
- outpun layer = num_class (1 karena hanya menentukan 1/0 kita bisa menggunakan output >= 0.5 sebagai 1 dan <0.5 sebagai 0)

In [ ]:
hist = model1c.fit(X_train, y_train, epochs=10, validation_data=(X_val, y_val))
loss, acc1c = model1c.evaluate(X_test, y_test)
print('\nAccuracy 1c:', acc1c)

Epoch 1/10
34/34 [==============================] - 1s 10ms/step - loss: 0.7337 - accuracy: 0.3935 - val_loss: 0.7248 - val_accuracy: 0.4478
Epoch 2/10
34/34 [==============================] - 0s 3ms/step - loss: 0.7124 - accuracy: 0.4318 - val_loss: 0.7061 - val_accuracy: 0.4627
Epoch 3/10
34/34 [==============================] - 0s 3ms/step - loss: 0.6987 - accuracy: 0.4692 - val_loss: 0.6940 - val_accuracy: 0.5299
Epoch 4/10
34/34 [==============================] - 0s 3ms/step - loss: 0.6900 - accuracy: 0.5168 - val_loss: 0.6864 - val_accuracy: 0.5373
Epoch 5/10
34/34 [==============================] - 0s 3ms/step - loss: 0.6837 - accuracy: 0.5336 - val_loss: 0.6810 - val_accuracy: 0.5448
Epoch 6/10
34/34 [==============================] - 0s 3ms/step - loss: 0.6788 - accuracy: 0.5551 - val_loss: 0.6769 - val_accuracy: 0.5672
Epoch 7/10
34/34 [==============================] - 0s 3ms/step - loss: 0.6747 - accuracy: 0.5626 - val_loss: 0.6733 - val_accuracy: 0.5970
Epoch 8/10
34/34 [=

Dengan menggunakan informasi dan hyperparameter dari soal dan default(sgd) sehingga diperoleh nilai accuracy seperti diatas.


Tunning hyperparameter:<br>
hyperparameters in neural networks: Number of layers, Number of neurons, Activation function, Learning rate,  Epochs, Optimizer, mommentum, and so on.<br><br>
Dan pada 1d saya melakukan tunning hyperparameter pada lr, momentum,number of layer, number of neurons,epoch.

In [ ]:
from keras.optimizers import SGD

n = 8

num_class = 1
sgd_optimizer = SGD(lr=0.1, momentum=0.9)
model1d = keras.models.Sequential(name="my_custom_model_1d")
model1d.add(tf.keras.Input(shape=(n,)))
model1d.add(keras.layers.core.Dense(2*n*5, activation='relu', name="Hiddenlayer"))
model1d.add(keras.layers.core.Dense(2*n*5,   activation='relu', name="Hiddenlayer2"))
model1d.add(keras.layers.core.Dense(2*n*5,   activation='relu', name="Hiddenlayer3"))
model1d.add(keras.layers.core.Dense(num_class,   activation='sigmoid', name="Outputlayer"))
model1d.compile(optimizer=sgd_optimizer, loss='binary_crossentropy', metrics=['accuracy'])

print(model1d.summary())

Model: "my_custom_model_1d"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 Hiddenlayer (Dense)         (None, 80)                720       
                                                                 
 Hiddenlayer2 (Dense)        (None, 80)                6480      
                                                                 
 Hiddenlayer3 (Dense)        (None, 80)                6480      
                                                                 
 Outputlayer (Dense)         (None, 1)                 81        
                                                                 
Total params: 13,761
Trainable params: 13,761
Non-trainable params: 0
_________________________________________________________________
None


/usr/local/lib/python3.10/dist-packages/keras/optimizers/legacy/gradient_descent.py:114: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super().__init__(name, **kwargs)


In [ ]:
hist2 = model1d.fit(X_train, y_train, epochs=16, validation_data=(X_val, y_val))
loss, acc1d = model1d.evaluate(X_test, y_test)

print('\nAccuracy 1c:', acc1c)
print('\nAccuracy 1d:', acc1d)

Epoch 1/16
34/34 [==============================] - 1s 8ms/step - loss: 0.6075 - accuracy: 0.6832 - val_loss: 0.6739 - val_accuracy: 0.6119
Epoch 2/16
34/34 [==============================] - 0s 3ms/step - loss: 0.5092 - accuracy: 0.7383 - val_loss: 0.3963 - val_accuracy: 0.8284
Epoch 3/16
34/34 [==============================] - 0s 3ms/step - loss: 0.3941 - accuracy: 0.8355 - val_loss: 0.4637 - val_accuracy: 0.7836
Epoch 4/16
34/34 [==============================] - 0s 3ms/step - loss: 0.4036 - accuracy: 0.8224 - val_loss: 0.3473 - val_accuracy: 0.8731
Epoch 5/16
34/34 [==============================] - 0s 3ms/step - loss: 0.3398 - accuracy: 0.8589 - val_loss: 0.3321 - val_accuracy: 0.8507
Epoch 6/16
34/34 [==============================] - 0s 3ms/step - loss: 0.3761 - accuracy: 0.8505 - val_loss: 0.3417 - val_accuracy: 0.8806
Epoch 7/16
34/34 [==============================] - 0s 3ms/step - loss: 0.3501 - accuracy: 0.8607 - val_loss: 0.3786 - val_accuracy: 0.8209
Epoch 8/16
34/34 [==

Setelah melakukan tunning hasil accuracy model semakin meningkat dari accuracy sebelum dilakukan tunning

Evaluasi MENGGUNAKAN PRECISSION, RECALL, F1-SCORE PADA NOMOR 1C

In [ ]:
model1c.compile(optimizer='sgd', loss='binary_crossentropy', metrics=[tf.keras.metrics.Precision(), tf.keras.metrics.Recall()])

In [ ]:
hist3 = model1c.fit(X_train, y_train, epochs=10, validation_data=(X_val, y_val))

loss, prec1c, rec1c = model1c.evaluate(X_test, y_test)
print('\nAccuracy:', acc1c)
print('F1-score:', 2*(prec1c*rec1c)/(prec1c+rec1c))
print('Precision:', prec1c)
print('Recall:', rec1c)

Epoch 1/10
34/34 [==============================] - 0s 4ms/step - loss: 0.6223 - precision_1: 0.5862 - recall_1: 1.0000 - val_loss: 0.6205 - val_precision_1: 0.6045 - val_recall_1: 1.0000
Epoch 2/10
34/34 [==============================] - 0s 3ms/step - loss: 0.6173 - precision_1: 0.5881 - recall_1: 0.9984 - val_loss: 0.6154 - val_precision_1: 0.6183 - val_recall_1: 1.0000
Epoch 3/10
34/34 [==============================] - 0s 3ms/step - loss: 0.6123 - precision_1: 0.5959 - recall_1: 0.9936 - val_loss: 0.6105 - val_precision_1: 0.6299 - val_recall_1: 0.9877
Epoch 4/10
34/34 [==============================] - 0s 3ms/step - loss: 0.6076 - precision_1: 0.6010 - recall_1: 0.9855 - val_loss: 0.6057 - val_precision_1: 0.6475 - val_recall_1: 0.9753
Epoch 5/10
34/34 [==============================] - 0s 3ms/step - loss: 0.6024 - precision_1: 0.6137 - recall_1: 0.9678 - val_loss: 0.6010 - val_precision_1: 0.6529 - val_recall_1: 0.9753
Epoch 6/10
34/34 [==============================] - 0s 3ms/s

Penjelasan mengenai hasil evaluasi dengan menggunakan metrics nilai accuracy, precision, recall dan F1-Score pada model1c:<br>
- Akurasi adalah rasio antara jumlah prediksi yang benar dan jumlah total prediksi. Nilai akurasi yang tinggi menunjukkan bahwa model dapat mengklasifikasikan data dengan tepat. Nilai akurasi 0.597 berarti bahwa model dapat mengklasifikasikan sekitar 60% data dengan benar, dan salah sekitar 40%. Ini menunjukkan bahwa model memiliki performa yang kurang baik dan perlu ditingkatkan.
- F1-score adalah rata-rata harmonik dari presisi dan recall. Nilai f1-score yang tinggi menunjukkan bahwa model memiliki keseimbangan yang baik antara presisi dan recall. Nilai f1-score 0.743 berarti bahwa model memiliki keseimbangan yang cukup baik antara presisi dan recall, tetapi masih ada ruang untuk perbaikan.
- Presisi adalah rasio antara jumlah prediksi positif yang benar dan jumlah total prediksi positif. Nilai presisi yang tinggi menunjukkan bahwa model dapat menghindari kesalahan (false positive). Nilai presisi 0.6 berarti bahwa model dapat menghindari sekitar 66% kesalahan positif palsu, tetapi masih membuat sekitar 34% kesalahan (false positive).
- Recall adalah rasio antara jumlah prediksi positif yang benar dan jumlah total data positif. Nilai recall yang tinggi menunjukkan bahwa model dapat mengenali sebagian besar data positif. Nilai recall 0.8500 berarti bahwa model dapat mengenali sekitar 85% data positif, tetapi masih melewatkan sekitar 15% data positif.

Dari hasil di atas, dapat disimpulkan bahwa model memiliki performa yang kurang baik dalam hal akurasi, tetapi memiliki keseimbangan yang cukup baik antara presisi dan recall. Model cenderung lebih sensitif dalam mengenali data positif daripada menghindari kesalahan (false positive).

EVALUASI MENGGUNAKAN PRECISSION, RECALL, F1-SCORE PADA NOMOR 1D

In [ ]:

model1d.compile(optimizer=sgd_optimizer, loss='binary_crossentropy', metrics=[tf.keras.metrics.Precision(), tf.keras.metrics.Recall()])


In [ ]:
hist4 = model1d.fit(X_train, y_train, epochs=16, validation_data=(X_val, y_val))

loss, prec1d, rec1d = model1d.evaluate(X_test, y_test)
print('\nAccuracy:', acc1d)
print('F1-score:', 2*(prec1d*rec1d)/(prec1d+rec1d))
print('Precision:', prec1d)
print('Recall:', rec1d)

Epoch 1/16
34/34 [==============================] - 1s 14ms/step - loss: 0.2836 - precision_2: 0.9199 - recall_2: 0.8682 - val_loss: 0.2524 - val_precision_2: 0.9589 - val_recall_2: 0.8642
Epoch 2/16
34/34 [==============================] - 0s 4ms/step - loss: 0.2802 - precision_2: 0.9021 - recall_2: 0.8891 - val_loss: 0.3124 - val_precision_2: 0.9697 - val_recall_2: 0.7901
Epoch 3/16
34/34 [==============================] - 0s 4ms/step - loss: 0.2537 - precision_2: 0.9165 - recall_2: 0.9003 - val_loss: 0.3262 - val_precision_2: 0.9545 - val_recall_2: 0.7778
Epoch 4/16
34/34 [==============================] - 0s 4ms/step - loss: 0.2789 - precision_2: 0.9136 - recall_2: 0.8666 - val_loss: 0.2726 - val_precision_2: 0.9559 - val_recall_2: 0.8025
Epoch 5/16
34/34 [==============================] - 0s 5ms/step - loss: 0.2536 - precision_2: 0.9168 - recall_2: 0.8859 - val_loss: 0.2844 - val_precision_2: 0.9114 - val_recall_2: 0.8889
Epoch 6/16
34/34 [==============================] - 0s 5ms/

Penjelasan mengenai hasil evaluasi dengan menggunakan metrics nilai accuracy, precision, recall dan F1-Score pada model1d:
Untuk menjelaskan hasil ini, Anda dapat menggunakan definisi dan interpretasi yang sama dengan sebelumnya. Misalnya, Anda dapat menjelaskan seperti ini:

- Akurasi adalah rasio antara jumlah prediksi yang benar dan jumlah total prediksi. Nilai akurasi yang tinggi menunjukkan bahwa model dapat mengklasifikasikan data dengan tepat. Nilai akurasi 0.88 berarti bahwa model dapat mengklasifikasikan mendekati 90% data dengan benar, dan salah sekitar 10%. Ini menunjukkan bahwa model memiliki performa yang baik dan akurat.
- F1-score adalah rata-rata harmonik dari presisi dan recall. Nilai f1-score yang tinggi menunjukkan bahwa model memiliki keseimbangan yang baik antara presisi dan recall. Nilai f1-score mendekati 0.9 berarti bahwa model memiliki keseimbangan yang sangat baik antara presisi dan recall, dan mendekati nilai optimal 1.
- Presisi adalah rasio antara jumlah prediksi positif yang benar dan jumlah total prediksi positif. Nilai presisi yang tinggi menunjukkan bahwa model dapat menghindari kesalahan positif palsu (false positive). Nilai presisi 0.9221 berarti bahwa model dapat menghindari sekitar 97% kesalahan (false positive), dan hanya membuat sekitar 3% kesalahan (false positive).
- Recall adalah rasio antara jumlah prediksi positif yang benar dan jumlah total data positif. Nilai recall yang tinggi menunjukkan bahwa model dapat mengenali sebagian besar data positif. Nilai recall 0.825 berarti bahwa model dapat mengenali sekitar 89% data positif, dan hanya melewatkan sekitar 18% data positif.

Dari hasil di atas, dapat disimpulkan bahwa model memiliki performa yang sangat baik dalam hal akurasi, presisi, recall, dan f1-score. Model cenderung lebih spesifik dalam menghindari kesalahan (false positive) daripada sensitif dalam mengenali data positif.

